### ISSUE TITLE
Traffic direction  in ntop is broken after 5.1.220130 update

### ISSUE THREAD

**Issue by notyourcommy** [ntop/ntopng#6267]
**Environment**:
* OS name: Centos 8
* OS version: centos 8 stream
* Architecture: adm65
* ntopng version/revision: 5.1.220130

**What happened**:
After last update (5.1.220130)  we see that in lua/if_stats.lua (interface statistics) traffic direction is broken
We are using local network configuration in both nprobe and ntopng (-L in nrprobe and -m in ntopng)

Also we see that now Main: Top Local Talkers and Main: Top Remote Talkers show only 5 ips out of 6 fields.
Ntopng sees our traffic as mostly outgoing now (which is false). It still sees some traffic as incoming
On the screenshot you can see before and after update
 
![ntop_after_update](https://user-images.githubusercontent.com/56963978/151810479-61ae339c-247e-49dd-8ee9-24f5998803af.png)

**Comment by MatteoBiscosi** [comment]
Hi @alexiff I fixed the top local/remote talkers limited to 5 instead of 6 (in a couple of hours at most a new update should be available). Instead regarding the traffic direction I tested on a machine but it's working correctly in my lab, still investigating the issue.
![Screenshot from 2022-01-31 16-27-58](https://user-images.githubusercontent.com/49585191/151821891-b1951942-c203-42b0-bc0a-620212a123f7.png)
(Latest ntopng 5.1.220131)

**Comment by notyourcommy** [comment]
> Hi @alexiff I fixed the top local/remote talkers limited to 5 instead of 6 (in a couple of hours at most a new update should be available). Instead regarding the traffic direction I tested on a machine but it's working correctly in my lab, still investigating the issue. ![Screenshot from 2022-01-31 16-27-58](https://user-images.githubusercontent.com/49585191/151821891-b1951942-c203-42b0-bc0a-620212a123f7.png) (Latest ntopng 5.1.220131)

Additional interesting info - if I remove "-L" from nprobe config ntopng thinks that all my traffic (well, 90% of my traffic) is now Incoming
![ntop-soso](https://user-images.githubusercontent.com/56963978/151822513-021d2ced-334c-43c6-a2bf-7b1552dfaa2e.png)

**Comment by lucaderi** [comment]
@alexiff We have reworked the wait direction work. Currenty
- if the direction is in the flow ntopng will honour it
- if the direction is not in the flow, we'll assume download remote->local host
Can you please share you configuration and how to reproduce the problem?

**Comment by notyourcommy** [comment]
@lucaderi 

What we have - 40gbit of traffic, radio is 50/50 (so 20gbit/20gbit). Traffic is exported only via sFlow.

1) No local networks - Result 35gbit incoming / 5gbit outgoing. No top local/remote talkers. Ratio is Wrong 

Nprobe
-i=none
--collector-port=6343
--zmq=tcp://127.0.0.1:5556
--zmq-probe-mode
-n=none
-G=/var/run/nprobe-sflow.pid
-T="%IN_SRC_MAC %OUT_DST_MAC %INPUT_SNMP %OUTPUT_SNMP %SRC_VLAN %IPV4_SRC_ADDR %IPV4_DST_ADDR %L4_SRC_PORT %L4_DST_PORT %IPV6_SRC_ADDR %IPV6_DST_ADDR %IP_PROTOCOL_VERSION %PROTOCOL %L7_PROTO %IN_BYTES %IN_PKTS %OUT_BYTES %OUT_PKTS %FIRST_SWITCHED %LAST_SWITCHED %CLIENT_TCP_FLAGS %SERVER_TCP_FLAGS %L7_PROTO_RISK %L7_RISK_SCORE %EXPORTER_IPV4_ADDRESS %DIRECTION%POST_NAT_SRC_IPV4_ADDR %POST_NAT_DST_IPV4_ADDR %POST_NAPT_SRC_TRANSPORT_PORT %POST_NAPT_DST_TRANSPORT_PORT %BGP_PREV_ADJACENT_ASN %BGP_NEXT_ADJACENT_ASN %IPV4_BGP_NEXT_HOP %SRC_AS %DST_AS"

Ntopng

-G=/var/run/ntopng.pid
-i "tcp://127.0.0.1:5556c"
-x 290000
-X 290000
--ignore-macs
--ignore-vlans
-F="clickhouse;127.0.0.1;ntopng;ntopng;######"
-v 2


2) Local networks in Nprobe only -  Result - Correct ratio ~18gig inc / ~22 out. But NO top remote/local talkers in NTOP.

Nprobe
-i=none
--collector-port=6343
--zmq=tcp://127.0.0.1:5556
--zmq-probe-mode
-n=none


-G=/var/run/nprobe-sflow.pid
-T="%IN_SRC_MAC %OUT_DST_MAC %INPUT_SNMP %OUTPUT_SNMP %SRC_VLAN %IPV4_SRC_ADDR %IPV4_DST_ADDR %L4_SRC_PORT %L4_DST_PORT %IPV6_SRC_ADDR %IPV6_DST_ADDR %IP_PROTOCOL_VERSION %PROTOCOL %L7_PROTO %IN_BYTES %IN_PKTS %OUT_BYTES %OUT_PKTS %FIRST_SWITCHED %LAST_SWITCHED %CLIENT_TCP_FLAGS %SERVER_TCP_FLAGS %L7_PROTO_RISK %L7_RISK_SCORE %EXPORTER_IPV4_ADDRESS %DIRECTION%POST_NAT_SRC_IPV4_ADDR %POST_NAT_DST_IPV4_ADDR %POST_NAPT_SRC_TRANSPORT_PORT %POST_NAPT_DST_TRANSPORT_PORT %BGP_PREV_ADJACENT_ASN %BGP_NEXT_ADJACENT_ASN %IPV4_BGP_NEXT_HOP %SRC_AS %DST_AS"
-L "subnet1,subnet2,subnet3"

Ntop config is the same as in (1)

3) Local networks in Ntopng only  - Result - traffic ratio as in (1) - like there is no configuration of local networks (almost everything is incoming) but now we see Top local/remote (but they also broken because we see same IPS in local and in remote)
Ntopng tells us that he knows about local networks

feb 01 14:58:09 novalocal ntopng[1347346]: 01/Feb/2022 14:58:09 [Ntop.cpp:3258] Added Local Network ####subnet1
feb 01 14:58:09 novalocal ntopng[1347346]: 01/Feb/2022 14:58:09 [Ntop.cpp:3258] Added Local Network ####subnet2
feb 01 14:58:09 novalocal ntopng[1347346]: 01/Feb/2022 14:58:09 [Ntop.cpp:3258] Added Local Network ####subnet3


-G=/var/run/ntopng.pid
-i "tcp://127.0.0.1:5556c"
-x 290000
-X 290000
--ignore-macs
--ignore-vlans
-F="clickhouse;127.0.0.1;ntopng;ntopng;######"
-v 2
-m "subnet1,subnet2,subnet3"

Nprobe config is the same as in (1)



Conclusion - ntopng option "-m"  is broken. Ntopng thinks that all traffic is still Incoming  and he puts local ips in  both top remote destinations  and top local talkers (our top local talkers and top remote destinations consist of SAME IP addresses)

p.s. - forgot situation 4) - both nprobe and ntop have local networks - result - Ratio now 90% outgoing. Broken Top Local/Remote as in 3. 
![ntop-cmoncmon](https://user-images.githubusercontent.com/56963978/151969465-6498f8f7-1c4e-430e-b6b1-570d090ea261.png)

**Comment by alexiff** [comment]
@MatteoBiscosi Nope, still top5

**Comment by lucaderi** [comment]
Some clarification. When you say the ratio is wrong you mean that this chart
![Screen Shot 2022-02-02 at 11 38 29](https://user-images.githubusercontent.com/4493366/152138310-02c60668-63d7-4efe-b4fc-8eee015a0d93.png)
is accurate as total but it is wrong in terms of % up/download. Correct? 

The top talkers issue was a bug that is now fixed and that will be included in the next build later today.

In sFlow we do not know the overall traffic direction as we know for each individual packet sample the in and out interface. So in order to merge the flows into something meaningful (sort of meaningful as if you monitor a switch you might see the same traffic multiple times in/out to make all this even more complex) we had to mimic direction with some tricks. We did it using the concept of local hosts as described here https://github.com/ntop/ntopng/blob/dev/src/ParserInterface.cpp#L497

In essence it works this way:
- ntopng
if we see traffic with no direction (so nProbe does not have -l)  from a local host -> remote host we assume this traffic is TX, otherwise is RX.

- nProbe
if nProbe collect flows with no direction (as with sFlow) it will mark flows with "unknown direction" leaving to ntopng the duty to mark them with the direction.

In summary
- set the local host in ntopng (suggested) or nProbe, but not in both
- wait for the new build available later today and let us know

**Comment by notyourcommy** [comment]
> is accurate as total but it is wrong in terms of % up/download. Correct?

Correct. 

What I did now
1) removed -L from nprobe
2) added -m to ntopng
3) updated to ntopng
Version:        5.1.220202 [Enterprise/Professional build]
GIT rev:        dev:0da4c556e77b2de9beba27c39b2c623e219fbba5:20220202
Pro rev:        r4739
I believe that is not the build you were talking about.

At the moment result is the same - top local list ips = ips in top remote.
Almost all traffic is RX

**Comment by notyourcommy** [comment]
Just updated ntop. Now top local/remote talkers/destinations are showing correct ips. Thank you.
Last thing is traffic direction being identified as all RX )

**Comment by notyourcommy** [comment]
@lucaderi Traffic direction in if_stats.lua is still wrong (just a reminder)

**Comment by lucaderi** [comment]
Sorry for the delay. The direction is set with -m unless your flow contains the direction. As you are using sFlow this could be the problem. How do you start nProbe exactly?

**Comment by notyourcommy** [comment]
Hello
Problem with direction is ONLY in if_stats.lua. 

Nprobe 
-i=none
-n=none
--zmq "tcp://127.0.0.1:5556"
--collector-port 6343
-V10
-b1
-T="%IPV4_SRC_ADDR %IPV4_DST_ADDR %L4_SRC_PORT %L4_DST_PORT %IPV6_SRC_ADDR %IPV6_DST_ADDR %IP_PROTOCOL_VERSION %SRC_TOS %PROTOCOL %ICMP_TYPE %INPUT_SNMP %SRC_AS %DST_AS %IPV4_NEXT_HOP %IPV6_NEXT_HOP %TCP_FLAGS %OUTPUT_SNMP %IN_BYTES %IN_PKTS %OUT_BYTES %OUT_PKTS %MIN_TTL %MAX_TTL %FIRST_SWITCHED %LAST_SWITCHED %SRC_VLAN %DST_VLAN %DOT1Q_SRC_VLAN %DOT1Q_DST_VLAN %EXPORTER_IPV4_ADDRESS %IN_SRC_MAC %IN_DST_MAC"

Ntop
-G=/var/run/ntopng.pid
-i "tcp://127.0.0.1:5556c"
-n 3
-x 290000
-X 290000
--ignore-vlans
-F="clickhouse;127.0.0.1;ntopng;ntopng;midiland"
-v 2 
-m "net1,net2,net3"

**Comment by MatteoBiscosi** [comment]
Hi @notyourcommy could which is your ntopng --version? could you please update and let me know if there is still this issue or not.

**Comment by MatteoBiscosi** [comment]
Closing for inactivity, please reopen if needed

**Comment by notyourcommy** [comment]
Can we open this issue again please ? Because it's a problem for many of ntopng users (some of them are even trying to bill their clients according to wrong interface statistics)

Issue : for some reason ntopng interface statistics treats most of the data as Incoming. (lua/if_stats.lua)

Our current ntop version is : 
Version:        5.7.230818 [Enterprise/Professional build]

cat /etc/ntopng/ntopng.conf
-G=/var/run/ntopng.pid
-F="clickhouse;*******;ntopng;ntop;******"
-i "tcp://127.0.0.1:5556"
-n 3
-x 700000
-X 700000
--ignore-vlans
-v 2 
-m "***********"

cat /etc/nprobe/nprobe-sflow.conf 
-i none
--collector-port 6343
--ntopng "zmq://127.0.0.1:5556"
-–flow-deduplication 15
-n none
-G=/var/run/nprobe-sflow.pid
-T="%IN_SRC_MAC %OUT_DST_MAC %INPUT_SNMP %OUTPUT_SNMP %SRC_VLAN %IPV4_SRC_ADDR %IPV4_DST_ADDR %L4_SRC_PORT %L4_DST_PORT %IPV6_SRC_ADDR %IPV6_DST_ADDR %IP_PROTOCOL_VERSION %PROTOCOL %L7_PROTO %IN_BYTES %IN_PKTS %OUT_BYTES %OUT_PKTS %FIRST_SWITCHED %LAST_SWITCHED %CLIENT_TCP_FLAGS %SERVER_TCP_FLAGS %L7_PROTO_RISK %L7_RISK_SCORE %EXPORTER_IPV4_ADDRESS %DIRECTION%POST_NAT_SRC_IPV4_ADDR %POST_NAT_DST_IPV4_ADDR %POST_NAPT_SRC_TRANSPORT_PORT %POST_NAPT_DST_TRANSPORT_PORT %BGP_PREV_ADJACENT_ASN %BGP_NEXT_ADJACENT_ASN %IPV4_BGP_NEXT_HOP %SRC_AS %DST_AS"


Here is correct graph from LibreNMS (snmp from interface)
![lnms-traffic](https://github.com/ntop/ntopng/assets/56963978/56edd4cb-ddb7-473f-a3f6-883e93ce7fd4)

And here is broken Ntop interface graph for the same time ( lua/if_stats.lua ) 

![ntop-traffic](https://github.com/ntop/ntopng/assets/56963978/eb726ac0-7b48-4982-927d-292a9fd11bd2)

**Comment by notyourcommy** [comment]
p.s. it doesn't matter what is in the "-m" in ntopng.conf, it doesnt matter if it's main interface or Disaggregated - ratio will always be strange like this (>80% incoming and ~20% outgoing).

Another proof that only lua/if_stats.lua is "broken" :
Same data as above but from Report : lua/pro/report.lua



![local-remote](https://github.com/ntop/ntopng/assets/56963978/8f47ee75-b183-44db-a034-57c4dded36fe)
You can see that it's like in LibreNMS screenshot from the above.

**Comment by lucaderi** [comment]
Fixed in nProbe: new overnight builds (dev) will contain the fix.
